### ISSUE TITLE
Support AQ

### ISSUE THREAD

**Issue by sinys15** [GoodManWEN/cx_Oracle_async#1]
Hello,
Could you please support AQ methods
https://cx-oracle.readthedocs.io/en/latest/user_guide/aq.html

**Comment by GoodManWEN** [comment]
Hi , @sinys15 

Since I ve been using mysql , I'm not quite familiar with what aq do in oracle. Maybe you can add yourself , welcoming pull request.

**Comment by ydf0509** [comment]
你在函数内部打印一下 threading.get_ident()，使不同的线程。我么我的感觉会是这样封装的asyncio的，效率还不如直接线程池了，loop.run_in_executor 这样做好处是什么？

**Comment by ydf0509** [comment]
异步并没有那么牛，实例太舍不得开线程数量了，什么不懂装懂的严格遵循cpu核心数的2n + 1，开多了线程生怕电脑像原子弹一样爆炸了一样。只要你大胆，线程池干趴asyncio，加qq来告诉你测试代码样本对比。

**Comment by GoodManWEN** [comment]
I'm going to close this issue.

**Comment by GoodManWEN** [comment]
AQ support was added in the latest branch , this is not a implement fully ported over functions from the original version. Different from official doc , modify AQ options is not supported , and the queue.deqMany() is not using original functions (for compatibility of  iterations in asynchronous code)

**Comment by sinys15** [comment]
Looks great but AQ options is absolutely required things.

**Comment by GoodManWEN** [comment]
Problem is it's not clear to me is the complexity and which of them will block the main loop.

**Comment by sinys15** [comment]
Could you please show the problem code.

**Comment by GoodManWEN** [comment]
There's nothing done yet no problem code. The problem to implement this is like we've got a [wrapper](https://github.com/GoodManWEN/cx_Oracle_async/blob/main/cx_Oracle_async/AQ.py#L3) of queue to make blocking io asynchronous , if we would like to set options at the wrapper level ,there should be a set of full-matched apis copied
from [dequeue-options](https://cx-oracle.readthedocs.io/en/latest/api_manual/aq.html#dequeue-options) and [enqueue-options](https://cx-oracle.readthedocs.io/en/latest/api_manual/aq.html#enqueue-options). That would be a whole bunche of work to separate those who will not block IO and those will. Besides , not sure if those option apis needs a thread block in threading mode , not much info found from cx-Oracle's doc.

**Comment by sinys15** [comment]
Is it a way to use both libraryes? cx_oracle for constract options and cx_oracle_async for execute enq/deq operation in async mode.

**Comment by GoodManWEN** [comment]
Perhaps that will do , then the remaining problem should be checkout if cx_oracle AQ options will or will not block IO , it's written in 
 C so its not easy to find out.

**Comment by sinys15** [comment]
AQ options is just the settings, they do nothing, it's like collections of settings compound in one object. The real work process in enq/deq methods and we need to set this object of options for this operations.

**Comment by GoodManWEN** [comment]
Oh, that's great , I'll draft a new develop branch soon. Still , I want to do some kind of test to make sure AQ operations do not block the IO ,and make sure the whole system is stable , yet currently no idea how to test it .

**Comment by GoodManWEN** [comment]
New branch drafted [dev](https://github.com/GoodManWEN/cx_Oracle_async/tree/882ff57914e2629aff4b740b18ad5c742ae7636f) . Maybe you can checkout if everything works fine , I'll then merge them to main branch.

**Comment by sinys15** [comment]
I have no idea why you make it so complicated way
```
async def modify_deqopts(queue , val):
    queue.deqOptions.wait = val
...
        queue = await conn.queue("DEMO_RAW_QUEUE")
        loop.create_task(modify_deqopts(queue , cx_Oracle.DEQ_NO_WAIT))
```

just enough
```
            queue = await conn.queue("DEMO_RAW_QUEUE")
            queue.enqOptions.visibility = cx_Oracle.ENQ_IMMEDIATE  # no need commit
            queue.deqOptions.visibility = cx_Oracle.ENQ_IMMEDIATE  # no need commit
            queue.deqOptions.wait = 2 # wait 2 sec
```
put settings is cheap and non blocking operation, you no need do it in new task.

And I don't understand why you make your own class DeqManyWrapper?

May be (I'm not sure) would be better make own enqMany by documentation warning
https://cx-oracle.readthedocs.io/en/latest/user_guide/aq.html#aqusermanual
Warning: calling Queue.enqMany() in parallel on different connections acquired from the same pool may fail due to Oracle bug 29928074. Ensure that this function is not run in parallel, use standalone connections or connections from different pools, or make multiple calls to Queue.enqOne() instead. The function Queue.deqMany() call is not affected.

**Comment by GoodManWEN** [comment]
Just a simulation of your changing options from somewhere else , but anyway , it's all running in the aio main loop , in another word the same thread , there should be no conflict at begining.

The reason why there's a DeqManyWrapper is cause that you need a object with `__aiter__` method to achieve `async for _ in queue.deqMany()` grammer , as the correspond usage of the synchronous version.
```Python
for m in queue.deqMany(maxMessages=10):
    print(m.payload.decode(connection.encoding))
```
In which they are using a generator.

I've already added a feature allows you to use the non-iterator grammer with deqMany like this:
```Python
async with oracle_pool.acquire() as conn:
    queue = await conn.queue('DEMO_RAW_QUEUE')
    results = await queue.deqMany()
    for  m in results:
        print(m.payload.decode(conn.encoding))
```

However the way of using a genrator added a benefit that you can take away all remaining objects from queue by not specifing maxMessages ,like this:
```Python
async with oracle_pool.acquire() as conn:
    queue = await conn.queue('DEMO_RAW_QUEUE')
    async for m in queue.deqMany()
        print(m.payload.decode(conn.encoding))
```
This will immediately return if there's a empty queue. And if youre using the non-iterator way , in which if there's a empty queue , the fetching thread will block forever until new object comes in , which may cause some not expected problem in async codes , that what I think.

**Comment by sinys15** [comment]
> I've already added a feature in the latest commit allows you to use the non-iterator grammer with deqMany like this:

I can't see your code but I hope it's use original cx_Oracle.deqMany inside.

**Comment by sinys15** [comment]
> This will immediately return if there's no longer anything in queue
It should be regulated by deq option NO_WAIT.

In my project I need dequeue in cycle forever and I do it inside asyncio.Task(my_func_with_deq_in_infinity_cycle)
I set dequeue timeout (to have opportunity to break program) and use while True cycle above it.

**Comment by sinys15** [comment]
My simple test works. I think you can release it. Anythink it don't touch old functionality.

**Comment by GoodManWEN** [comment]
0.2.1 released on pypi.

**Comment by GoodManWEN** [comment]
0.2.2 released on pypi. 

Find it really slow using `async for _ in queue.deqMany():` during 0.2.1 performance test , probably cause of the  communication time and there's a asyncio.Lock aquire and release action each query. So code rewrited and now `async for _ in queue.deqMany():` is using the same api as `await queue.deqMany()` , they are expect to have the same performance.
